engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java

```diff
@@ -41,64 +44,87 @@ boolean isEmpty() {
    * Adds an element to the builder
    *
    * @param e the element to add
+   * @param warnings library to check for values with warnings
    */
-  void add(Object e) {
-    if (objectArray != null) {
-      if (size == objectArray.length) {
-        CompilerDirectives.transferToInterpreter();
-        objectArray = Arrays.copyOf(objectArray, size * 2);
+  void add(Object e, WarningsLibrary warnings) {
+    if (!nonTrivialEnsoValue) {
+      if (warnings.hasWarnings(e)) {
+        nonTrivialEnsoValue = true;
+      } else {
+        var isEnsoValue = e instanceof EnsoObject || e instanceof Long || e instanceof Double;
+        if (!isEnsoValue) {
+          nonTrivialEnsoValue = true;
+        }
       }
-      objectArray[size++] = e;
-    } else if (primitiveArray instanceof long[] longArray) {
-      if (e instanceof Long l) {
-        if (size == longArray.length) {
+    }
+    addImpl(e);
+  }
+
+  private void addImpl(Object e) {
+    while (true) {
+      // loop at most twice thanks to continue and break
```

Comment:
I find this continue/break logic hard to follow.

I suggest breaking the logic into smaller functions, allowing much clearer control flow:
```diff
diff --git a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java b/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java
--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java	(revision 4dbb1506eb1344a8ebda2784fd7586cf3d1f1b3e)
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java	(date 1717144249779)
@@ -47,85 +47,98 @@
    * @param warnings library to check for values with warnings
    */
   void add(Object e, WarningsLibrary warnings) {
-    if (!nonTrivialEnsoValue) {
-      if (warnings.hasWarnings(e)) {
-        nonTrivialEnsoValue = true;
-      } else {
-        var isEnsoValue = e instanceof EnsoObject || e instanceof Long || e instanceof Double;
-        if (!isEnsoValue) {
-          nonTrivialEnsoValue = true;
-        }
-      }
-    }
-    addImpl(e);
-  }
+    updateNonTrivialValueCheck(e, warnings);
 
-  private void addImpl(Object e) {
-    while (true) {
-      // loop at most twice thanks to continue and break
-      if (objectArray != null) {
-        if (size == objectArray.length) {
-          CompilerDirectives.transferToInterpreter();
-          objectArray = Arrays.copyOf(objectArray, size * 2);
-        }
-        objectArray[size++] = e;
-      } else if (primitiveArray instanceof long[] longArray) {
-        if (e instanceof Long l) {
-          if (size == longArray.length) {
-            CompilerDirectives.transferToInterpreter();
-            primitiveArray = longArray = Arrays.copyOf(longArray, size * 2);
-          }
-          longArray[size++] = l;
-        } else {
-          CompilerDirectives.transferToInterpreter();
-          objectArray = new Object[longArray.length];
-          for (int i = 0; i < size; i++) {
-            objectArray[i] = longArray[i];
-          }
-          primitiveArray = null;
-          continue;
-        }
-      } else if (primitiveArray instanceof double[] doubleArray) {
-        if (e instanceof Double d) {
-          if (size == doubleArray.length) {
-            CompilerDirectives.transferToInterpreter();
-            primitiveArray = doubleArray = Arrays.copyOf(doubleArray, size * 2);
-          }
-          doubleArray[size++] = d;
-        } else {
-          CompilerDirectives.transferToInterpreter();
-          objectArray = new Object[doubleArray.length];
-          for (int i = 0; i < size; i++) {
-            objectArray[i] = doubleArray[i];
-          }
-          primitiveArray = null;
-          continue;
-        }
-      } else {
-        assert objectArray == null;
-        assert primitiveArray == null;
-        assert size == 0;
-        switch (e) {
-          case Long l -> {
-            var arr = new long[initialCapacity];
-            arr[0] = l;
-            primitiveArray = arr;
-          }
-          case Double d -> {
-            var arr = new double[initialCapacity];
-            arr[0] = d;
-            primitiveArray = arr;
-          }
-          default -> {
-            var arr = new Object[initialCapacity];
-            arr[0] = e;
-            objectArray = arr;
-          }
-        }
-        size = 1;
-      }
-      break;
-    }
+    if (objectArray != null) {
+      addObject(e);
+    } else if (primitiveArray instanceof long[] longArray) {
+      if (e instanceof Long l) {
+        if (size == longArray.length) {
+          CompilerDirectives.transferToInterpreter();
+          primitiveArray = longArray = Arrays.copyOf(longArray, size * 2);
+        }
+        longArray[size++] = l;
+      } else {
+        retypeToObject(longArray);
+        addObject(e);
+      }
+    } else if (primitiveArray instanceof double[] doubleArray) {
+      if (e instanceof Double d) {
+        if (size == doubleArray.length) {
+          CompilerDirectives.transferToInterpreter();
+          primitiveArray = doubleArray = Arrays.copyOf(doubleArray, size * 2);
+        }
+        doubleArray[size++] = d;
+      } else {
+        retypeToObject(doubleArray);
+        addObject(e);
+      }
+    } else {
+      // initialize with first element
+      assert objectArray == null;
+      assert primitiveArray == null;
+      assert size == 0;
+      switch (e) {
+        case Long l -> {
+          var arr = new long[initialCapacity];
+          arr[0] = l;
+          primitiveArray = arr;
+        }
+        case Double d -> {
+          var arr = new double[initialCapacity];
+          arr[0] = d;
+          primitiveArray = arr;
+        }
+        default -> {
+          var arr = new Object[initialCapacity];
+          arr[0] = e;
+          objectArray = arr;
+        }
+      }
+      size = 1;
+    }
+  }
+
+  private void updateNonTrivialValueCheck(Object e, WarningsLibrary warnings) {
+    if (!nonTrivialEnsoValue) {
+      if (warnings.hasWarnings(e)) {
+        nonTrivialEnsoValue = true;
+      } else {
+        var isTrivial = e instanceof EnsoObject || e instanceof Long || e instanceof Double;
+        if (!isTrivial) {
+          nonTrivialEnsoValue = true;
+        }
+      }
+    }
+  }
+
+  private void addObject(Object e) {
+    assert objectArray != null;
+    if (size == objectArray.length) {
+      CompilerDirectives.transferToInterpreter();
+      objectArray = Arrays.copyOf(objectArray, size * 2);
+    }
+    objectArray[size++] = e;
+  }
+
+  private void retypeToObject(long[] longArray) {
+    CompilerDirectives.transferToInterpreter();
+    assert objectArray == null;
+    objectArray = new Object[longArray.length];
+    for (int i = 0; i < size; i++) {
+      objectArray[i] = longArray[i];
+    }
+    primitiveArray = null;
+  }
+
+  private void retypeToObject(double[] longArray) {
+    CompilerDirectives.transferToInterpreter();
+    assert objectArray == null;
+    objectArray = new Object[longArray.length];
+    for (int i = 0; i < size; i++) {
+      objectArray[i] = longArray[i];
+    }
+    primitiveArray = null;
   }
 
   /** Obtains an element from the builder */
@@ -232,9 +245,11 @@
   Object asVector() {
     var res = toArray();
     if (res instanceof long[] longs) {
+      assert !nonTrivialEnsoValue;
       return Vector.fromLongArray(longs);
     }
     if (res instanceof double[] doubles) {
+      assert !nonTrivialEnsoValue;
       return Vector.fromDoubleArray(doubles);
     }
     if (nonTrivialEnsoValue) {
```

I think it is easier to understand if there is no loops.

Reply:
But my suggested solution does not have a recursive call. Shouldn't the compiler be able to inline `addObject`?

Reply:
`addObject` is a nice refactoring. Yes, inlining shall work fine. Applied in 52f815b